Products/DataCollector/zing/DatamapHandler.py

```diff
@@ -88,21 +94,36 @@ def process_datamaps(self, tx_success, zing_state):
         if not tx_success or not self.zing_connector_url:
             return
         try:
-            log.info("Sending {} datamaps to zing-connector.".format(len(zing_state.datamaps)))
+            log.debug("Sending {} datamaps to zing-connector.".format(len(zing_state.datamaps)))
+            facts = []
             for device, datamap in zing_state.datamaps:
-                # TODO should we batch them instead of sending them one by one?
-                self._send_datamap(device, datamap, zing_state.contexts)
+                dm_facts = facts_from_datamap(device, datamap, zing_state.contexts)
+                if dm_facts:
+                    facts.extend(dm_facts)
+            # add dummy facts for the processed devices
+            device_facts = zing_state.devices_fact.values()
+            if device_facts:
+                facts.extend(device_facts)
+            self._send_facts_in_batches(facts)
         except Exception as ex:
             log.warn("Exception sending datamaps to zing-connector. {}".format(ex))
 
-    def _send_datamap(self, device, datamap, context):
+    def _send_facts_in_batches(self, facts, batch_size=500):
```

Comment:
Does this also need an alternate time qualifier? For example, as in batch_size=500 || batch_time=300s . So the batch is sent when either of the size or time constraints are met

We need to handle the case wherein the number of facts are less than the batch size and so we are not waiting indefinitely (or rather, for the remodel interval ) to complete the batch size

Reply:
We always send all the facts once the tx is successfully committed

Reply:
Ok, that works